File: rust_vs_c++/heritage/heritage.rs
// Hugo BERANGER - M2 MIAGE IA

trait Forme {
    fn get_nom(&self) -> &'static str;
    fn who_am_i(&self) -> &'static str {
        self.get_nom()
    }
}

struct Rectangle {
    nom: &'static str,
}

struct Carre {
    nom: &'static str,
}

impl Forme for Rectangle {
    fn get_nom(&self) -> &'static str {
        &self.nom
    }
}

impl Forme for Carre {
    fn get_nom(&self) -> &'static str {
        &self.nom
    }
}

fn main() {
    let rectangle_objet = Rectangle {
        nom: "rectangle",
    };
    println!("rust  - rectangle nom : {}",rectangle_objet.who_am_i());

    let carre_objet = Carre {
        nom: "carré",
    };
    println!("rust  - carré nom : {}",carre_objet.who_am_i());
}

File: rust_vs_c++/heritage/heritage.cpp
// Hugo BERANGER - M2 MIAGE IA

#include <iostream>

using namespace std;

class Forme {
    private:
        string nom;
    public:
        Forme(string n){
            nom = n;
        };
        string who_am_i() const{
            return nom; 
        }
};

class Rectangle : public Forme {
    public:
        Rectangle (string n) : Forme(n){}
};

class Carre : public Forme {
    public:
        Carre (string n) : Forme(n){}
};

int main () {
    Rectangle rectangleObjet("rectangle");
    cout << "c++  - rectangle : " << rectangleObjet.who_am_i() << endl;
    Carre carreObjet("carré");
    cout << "c++  - carre : " << carreObjet.who_am_i() << endl;
    return 0;
}

